### PR TITLE
Add new methods to get External Workflow Identifier and Template id from Workflow Id

### PR DESCRIPTION
**Pull request by malithiperera**
## Implement a method to retrieve External Workflow Id  

The getExternalWorkflowId function retrieves the external workflow identifier associated with a given workflow id. This unique identifier corresponds to the deployed workflow id in an external BPS engine.

For an example external workflow identifier of  a workflow deployed in Camunda :

image

### CODE REVIEW

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java [hunk 1]
```diff
@@ -96,6 +95,10 @@ public interface WorkflowManagementService {
     void addWorkflow(Workflow workflowDTO,
                      List<Parameter> parameterList, int tenantId) throws WorkflowException;
 
+    String getTemplateId(String workflowId) throws  WorkflowException;
```
Comment (sadilchamishka) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:98]: Let's add java doc comments for the new methods. Check the above listed methods for reference.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java [hunk 2]
```diff
@@ -126,6 +126,54 @@ public WorkflowRequest retrieveWorkflow(String uuid) throws InternalWorkflowExce
         return null;
     }
 
+    public static String retrieveExternalWorkflowId(String workflowId) throws InternalWorkflowException {
```
Comment (sadilchamishka) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:129]: Remove the static modifier and add method comments. You can refer another method in this class.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 3]
```diff
@@ -383,6 +383,14 @@ public void addWorkflow(Workflow workflow,
 
     }
 
+    @Override
+    public String getTemplateId(String workflowId) throws WorkflowException {
+        return WorkflowRequestDAO.retrieveTemplateId(workflowId);
```
Comment (sadilchamishka) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java:388]: When the static modified of the methods are removed, the methods will have to be accessed as
`workflowDAO.retrieveTemplateId(workflowId);`

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java [hunk 4]
```diff
@@ -126,6 +126,54 @@ public WorkflowRequest retrieveWorkflow(String uuid) throws InternalWorkflowExce
...
+        String query = SQLConstants.GET_WORKFLOW_ID_QUERY;
+        try {
+            prepStmt = connection.prepareStatement(query);
+            prepStmt.setString(1, workflowId);
+            rs = prepStmt.executeQuery();
+            if (rs.next()) {
+                return rs.getString(SQLConstants.WORKFLOWID_COLUMN);
+            }
+        } catch (SQLException e) {
+
```
Comment (sadilchamishka) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:144]: Remove unnecessary new lines before and after line 145.

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java [hunk 5]
```diff
@@ -57,13 +57,18 @@ public class SQLConstants {
     public static final String REQUEST_ID_COLUMN = "REQUEST_ID";
     public static final String CREATED_BY_COLUMN = "CREATED_BY";
 
+    public static  final String WORKFLOWID_COLUMN = "PARAM_VALUE";
```
Comment (sadilchamishka) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java:60]: No need to define this constant. It i already defined. https://github.com/wso2/carbon-identity-framework/pull/4519/files#diff-6b77d285eb536d0c997683f1ea8dd0482bf4d6afbaeeebbf4ef2cc51f3e5dae3R45
Btw try to keep generic when defining a new constant. The variable should have been `PARAM_VALUE_COLUMN` instead of `WORKFLOWID_COLUMN`

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java [hunk 6]
```diff
@@ -57,13 +57,18 @@ public class SQLConstants {
...
 
     public static final String ADD_WORKFLOW_REQUEST_QUERY = "INSERT INTO WF_REQUEST(UUID, CREATED_BY, OPERATION_TYPE," +
             " CREATED_AT, UPDATED_AT, REQUEST, STATUS, TENANT_ID) VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
 
     public static final String GET_WORKFLOW_REQUEST_QUERY = "SELECT UUID, REQUEST, STATUS, CREATED_BY FROM WF_REQUEST" +
             " WHERE UUID = ?";
-
+    public static final String GET_WORKFLOW_ID_QUERY = "SELECT PARAM_VALUE FROM WF_WORKFLOW_CONFIG_PARAM WHERE " +
+            "WORKFLOW_ID  = ?" +
+            " AND PARAM_NAME ='ExternalWorkflowIdentifier'";
```
Comment (sadilchamishka) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java:70]: combine into single line

Comment (sadilchamishka) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java:70]: PARAM_NAME = 'ExternalWorkflowIdentifier'.  also let's keep a space character after `=` operator

File: components/application-mgt/org.wso2.carbon.identity.application.common/pom.xml [hunk 7]
```diff
@@ -14,7 +14,8 @@
   ~ See the License for the specific language governing permissions and
   ~ limitations under the License.
   -->
-<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
```
[the remark this answers is not in the document]
Comment (malithiperera) [reply components/application-mgt/org.wso2.carbon.identity.application.common/pom.xml]: Addressed the comment ref: https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 8]
```diff
@@ -383,6 +383,14 @@ public void addWorkflow(Workflow workflow,
...
+    }
+    public String getExternalWorkflowId(String workflowId) throws WorkflowException{
```
[the remark this answers is not in the document]
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java:393]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java (hunk 1 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:104]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java (hunk 4 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java (hunk 5 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java (hunk 6 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java [hunk 9]
```diff
@@ -57,13 +57,18 @@ public class SQLConstants {
     public static final String REQUEST_ID_COLUMN = "REQUEST_ID";
     public static final String CREATED_BY_COLUMN = "CREATED_BY";
 
+    public static  final String WORKFLOWID_COLUMN = "PARAM_VALUE";
+    public static final String TEMPLATEID_COLUMN = "TEMPLATE_ID";
```
[the remark this answers is not in the document]
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java]: addressed the comment ref:https://github.com/wso2/carbon-identity-framework/pull/4519/commits/ce9bfa08202eb691d2a4f87aebd30fe3445f0fd5

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java [hunk 10]
```diff
@@ -96,6 +95,22 @@ public interface WorkflowManagementService {
     void addWorkflow(Workflow workflowDTO,
                      List<Parameter> parameterList, int tenantId) throws WorkflowException;
 
+    /**
+     *Get Template ID from workflow Id
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:99]: add a space after the star and ends with full stop

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java [hunk 11]
```diff
@@ -96,6 +95,22 @@ public interface WorkflowManagementService {
...
+     * @param workflowId workflow ID retrieve from workflow request
+     * @return Template ID
+     * @throws WorkflowException
+     */
+    String getTemplateId(String workflowId) throws  WorkflowException;
+
+    /**
+     * Get external Workflow Identifier (Workflow Deploy ID)
+     * @param workflowId workflow ID retrieve from workflow request
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:108]: add a new line after dis and explain the return and all the parameters

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java [hunk 12]
```diff
@@ -96,6 +95,22 @@ public interface WorkflowManagementService {
...
+    /**
+     *Get Template ID from workflow Id
+     * @param workflowId workflow ID retrieve from workflow request
+     * @return Template ID
+     * @throws WorkflowException
+     */
+    String getTemplateId(String workflowId) throws  WorkflowException;
+
+    /**
+     * Get external Workflow Identifier (Workflow Deploy ID)
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:107]: get the identifier that is generated when an external workflow is deployed

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java [hunk 13]
```diff
@@ -73,7 +73,7 @@ public class WorkflowManagementServiceImpl implements WorkflowManagementService
     WorkflowDAO workflowDAO = new WorkflowDAO();
     AssociationDAO associationDAO = new AssociationDAO();
     private RequestEntityRelationshipDAO requestEntityRelationshipDAO = new RequestEntityRelationshipDAO();
-    private WorkflowRequestDAO workflowRequestDAO = new WorkflowRequestDAO();
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementServiceImpl.java:76]: revert this change

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java [hunk 14]
```diff
@@ -126,6 +126,68 @@ public WorkflowRequest retrieveWorkflow(String uuid) throws InternalWorkflowExce
...
+        try {
+            prepStmt = connection.prepareStatement(query);
+            prepStmt.setString(1, workflowId);
+            rs = prepStmt.executeQuery();
+            if (rs.next()) {
+                return rs.getString(SQLConstants.PARAM_VALUE_COLUMN);
+            }
+        } catch (SQLException e) {
+            throw new InternalWorkflowException("Error when executing the sql query:" + query, e);
+
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:152]: remove new line

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java [hunk 15]
```diff
@@ -126,6 +126,68 @@ public WorkflowRequest retrieveWorkflow(String uuid) throws InternalWorkflowExce
...
+        String query = SQLConstants.GET_WORKFLOW_ID_QUERY;
+        try {
+            prepStmt = connection.prepareStatement(query);
+            prepStmt.setString(1, workflowId);
+            rs = prepStmt.executeQuery();
+            if (rs.next()) {
+                return rs.getString(SQLConstants.PARAM_VALUE_COLUMN);
+            }
+        } catch (SQLException e) {
+            throw new InternalWorkflowException("Error when executing the sql query:" + query, e);
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:151]: improve this :  handleException(WFConstant.Exceptions.SQL_ERROR_GETTING_WORKFLOW_COUNT, e);

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java [hunk 16]
```diff
@@ -126,6 +126,68 @@ public WorkflowRequest retrieveWorkflow(String uuid) throws InternalWorkflowExce
...
     }
 
+    /**
+     * Retrieve the External Workflow Identifier from Workflow ID.
+     *
+     * @param workflowId workflow ID retrieve from workflow request
+     * @return External Workflow Identifier
+     * @throws InternalWorkflowException
+     */
+    public String retrieveExternalWorkflowId(String workflowId) throws InternalWorkflowException {
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:136]: return optional

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java [hunk 17]
```diff
@@ -126,6 +126,68 @@ public WorkflowRequest retrieveWorkflow(String uuid) throws InternalWorkflowExce
...
+            if (rs.next()) {
+                return rs.getString(SQLConstants.PARAM_VALUE_COLUMN);
+            }
+        } catch (SQLException e) {
+            throw new InternalWorkflowException("Error when executing the sql query:" + query, e);
+
+        } finally {
+            IdentityDatabaseUtil.closeAllConnections(connection, null, prepStmt);
+        }
+        return null;
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:156]: optional.nullable

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java [hunk 18]
```diff
@@ -57,13 +57,14 @@ public class SQLConstants {
...
     public static final String CREATED_BY_COLUMN = "CREATED_BY";
 
-
     public static final String ADD_WORKFLOW_REQUEST_QUERY = "INSERT INTO WF_REQUEST(UUID, CREATED_BY, OPERATION_TYPE," +
             " CREATED_AT, UPDATED_AT, REQUEST, STATUS, TENANT_ID) VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
 
     public static final String GET_WORKFLOW_REQUEST_QUERY = "SELECT UUID, REQUEST, STATUS, CREATED_BY FROM WF_REQUEST" +
             " WHERE UUID = ?";
-
+    public static final String GET_WORKFLOW_ID_QUERY = "SELECT PARAM_VALUE FROM WF_WORKFLOW_CONFIG_PARAM WHERE " +
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java:65]: find out Template value in database

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java [hunk 19]
```diff
@@ -57,13 +57,14 @@ public class SQLConstants {
...
+            "WORKFLOW_ID  = ? AND PARAM_NAME = 'ExternalWorkflowIdentifier'";
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java:66]: create a constant and pass it when the function call

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java [hunk 20]
```diff
@@ -57,13 +57,14 @@ public class SQLConstants {
     public static final String REQUEST_ID_COLUMN = "REQUEST_ID";
     public static final String CREATED_BY_COLUMN = "CREATED_BY";
 
-
     public static final String ADD_WORKFLOW_REQUEST_QUERY = "INSERT INTO WF_REQUEST(UUID, CREATED_BY, OPERATION_TYPE," +
             " CREATED_AT, UPDATED_AT, REQUEST, STATUS, TENANT_ID) VALUES (?, ?, ?, ?, ?, ?, ?, ?)";
 
     public static final String GET_WORKFLOW_REQUEST_QUERY = "SELECT UUID, REQUEST, STATUS, CREATED_BY FROM WF_REQUEST" +
             " WHERE UUID = ?";
-
```
Comment (malithiperera) [inline components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/util/SQLConstants.java:66]: revert the change

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java (hunk 11 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java:101]: fix by : https://github.com/wso2/carbon-identity-framework/pull/4519/commits/038a1a1ca8e9021e90440538fcf4251b989aa3e0

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java (hunk 14 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/dao/WorkflowRequestDAO.java:158]: fix by : https://github.com/wso2/carbon-identity-framework/pull/4519/commits/038a1a1ca8e9021e90440538fcf4251b989aa3e0

File: components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java (hunk 10 above)
Comment (malithiperera) [reply components/workflow-mgt/org.wso2.carbon.identity.workflow.mgt/src/main/java/org/wso2/carbon/identity/workflow/mgt/WorkflowManagementService.java]: Addressed
